pmemcheck/pmc_main.c

```diff
@@ -824,8 +965,21 @@ trace_pmem_store(Addr addr, SizeT size, UWord value)
 {
     if (LIKELY(!is_pmem_access(addr, size)))
         return;
-    struct pmem_st *store = VG_(OSetGen_AllocNode)(pmem.pmem_stores,
+    struct pmem_st *store;
+    UWord old_index=pmem.info_array.m_data[pmem.info_array.m_index].end_index;
+    if (LIKELY(old_index < MAX_ARRAY_NUM))
```

Comment:
Yes, we need to create two data structures for tracking stores, but persistent memory information doesn't pick one of them to insert.
Actually, persistent memory information must be inserted into the array structure at the first time (processing store instructions), and then they will transfer from the array to the tree structure (processing fence instructions).

Last, I don't think the abstraction is necessary because we have to process data from both structures.